src/sage/rings/species.py

```diff
@@ -0,0 +1,1468 @@
+            sage: A = At(G, {1: [1,2,3,4], 2: [5,6,7,8,9,10]}); A
+            {((1,2,3,4)(5,6)(7,8)(9,10),): (frozenset({8, 5, 6, 7}), frozenset({1, 2, 3, 4, 9, 10}))}
+            sage: B = At(H, {1: [1,2,3,4], 2: [5,6,7,8,9,10]}); B
+            {((1,2,3,4)(5,6)(7,8)(9,10),): (frozenset({1, 2, 3, 4}), frozenset({5, 6, 7, 8, 9, 10}))}
+            sage: C = At(G, {1: [1,2,5,6], 2: [3,4,7,8,9,10]}); C
+            {((1,2,3,4)(5,6)(7,8)(9,10),): (frozenset({8, 5, 6, 7}), frozenset({1, 2, 3, 4, 9, 10}))}
+            sage: hash(A) == hash(B)
+            True
+            sage: hash(A) == hash(C)
+            True
+        """
+        return hash(self._element_key())
+
+    def __eq__(self, other):
+        r"""
+        Two atomic species are equal if the underlying groups are conjugate,
+        and their domain partitions are equal under the conjugating element.
+
+        TESTS::
+
+            sage: At = AtomicSpecies("X, Y")
+            sage: G = PermutationGroup([[(1,2),(3,4),(5,6),(7,8,9,10)]]); G
+            Permutation Group with generators [(1,2)(3,4)(5,6)(7,8,9,10)]
+            sage: H = PermutationGroup([[(1,2,3,4),(5,6),(7,8),(9,10)]]); H
+            Permutation Group with generators [(1,2,3,4)(5,6)(7,8)(9,10)]
+            sage: A = At(G, {1: [1,2,3,4], 2: [5,6,7,8,9,10]})
+            sage: B = At(H, {1: [1,2,3,4], 2: [5,6,7,8,9,10]})
+            sage: C = At(G, {1: [1,2,5,6], 2: [3,4,7,8,9,10]})
+            sage: A != B
+            True
+            sage: A == C
+            True
+        """
+        if parent(self) != parent(other):
+            return False
+        # Check if multicardinalities match
+        if self._mc != other._mc:
+            return False
+        # If they do, construct the mapping between the groups
+        selflist, otherlist = [], []
+        for i in range(self.parent()._k):
+            if len(self._dompart[i]) != len(other._dompart[i]):
+                return False
+            selflist.extend(sorted(list(self._dompart[i])))
+            otherlist.extend(sorted(list(other._dompart[i])))
+        mapping = libgap.MappingPermListList(selflist, otherlist)
+        G = PermutationGroup(gap_group=libgap.ConjugateGroup(self._dis._C, mapping),
+                             domain=self._dis._C.domain())
+        # The conjugated group must be exactly equal to the other group
+        return G == other._dis._C
+
+    def _repr_(self):
+        r"""
+        Return a string representation of ``self``.
+
+        TESTS::
+
+            sage: At = AtomicSpecies("X, Y")
+            sage: G = PermutationGroup([[(1,2),(3,4),(5,6),(7,8,9,10)]]); G
+            Permutation Group with generators [(1,2)(3,4)(5,6)(7,8,9,10)]
+            sage: A = At(G, {1: [1,2,3,4], 2: [5,6,7,8,9,10]}); A
+            {((1,2,3,4)(5,6)(7,8)(9,10),): (frozenset({8, 5, 6, 7}), frozenset({1, 2, 3, 4, 9, 10}))}
+        """
+        return "{" + f"{self._dis}: {self._dompart}" + "}"
```

Comment:
```suggestion
        return "{" + f"{self._dis}: {SetPartition(self._dompart)}" + "}"
```
together with
```
from sage.combinat.set_partition import SetPartition
```
at the beginning.

Reply:
Sorry, that doesn't work (because of the empty sets), we should do the printing ourselves.

Maybe something like `", ".join(Set(b) for b in self._dompart)`